---
title: "RNA-seq differential expression testing"
author: "Sam Buckberry"
date: "30/10/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, message=FALSE}
library(magrittr)
library(stringr)
library(edgeR)
library(ggplot2)
library(pheatmap)
library(gprofiler2)
library(wiggleplotr)
```

Load the sample data
```{r}
mdat <- read.csv("metadata/sample_data.csv")
mdat
```

Load the gene and transcript counts data
```{r}
gene_counts <- readRDS("processed_data/Effector_CD4pos_ALL_gene_counts.Rds") 
tx_counts <- readRDS("processed_data/Effector_CD4pos_ALL_transcript_counts.Rds")
```

Ensure metadata and data labels are in the same order
```{r}
if (all(mdat$Bam != gene_counts$targets)){
    ind <- match(gene_counts$targets, mdat$Bam)
    mdat <- mdat[ind, ]
}

stopifnot(all(mdat$Bam == gene_counts$targets))
```


Create the DGElist object for using edgeR
```{r}
y <- DGEList(counts = gene_counts$counts,
             samples = mdat$Sample,
             group = mdat$Treatment)

```

### Design matrix for differential testing
```{r}
design <- model.matrix(~ 0 + y$samples$group)
design
# Clean up the column names (will make things easier later)
colnames(design) <- c("Stimulated", "Unstimulated")
design
```

Boxplot of CPM normalised expression for each sample
```{r}
boxplot(cpm(y, log=TRUE), names=y$samples$samples)
```


Filter genes/transcripts to remove genes with low counts that would not be useful in statistical testing
```{r}
keep <- filterByExpr(y)
table(keep)

y <- y[keep, , keep.lib.sizes=FALSE]
```

### Calculate normalisation factors
```{r}
y <- calcNormFactors(y)
y$samples
```

### Estimate dispersions
```{r}
y <- estimateDisp(y, robust = TRUE)
```

### MDS Plot
How do the samples group based on the genes with the largest differences? Do they group by sample donor or by treatment group? In this type of plot, you want to look for outliers. 

```{r}
plotMDS(y, labels = y$samples$samples)
plotMDS(y, labels = y$samples$group)
```

### Boxplots

Boxplots of log2 counts for each sample
```{r, fig.height=5}
boxplot(cpm(y, log=TRUE), names=y$samples$samples)
```

Differential testing using the exact test in edgeR
```{r}
et <- exactTest(y)
tt <- topTags(et, n = nrow(y))
```

How many DE genes are there with FDR < 0.05 and logFC > 1?
```{r}
sum((tt$table$FDR < 0.05) & (abs(tt$table$logFC) > 1))
```

### Differential testing using Limma-Voom
```{r}
v <- voom(counts = y, design = design, plot = TRUE)
```


```{r, fig.height=5}
boxplot(v$E, names=v$targets$samples)
```

If you want to normalise further using quantile normalisation
```{r}
vq <- voom(counts = y, design = design, plot = TRUE,
           normalize.method = "quantile")
```

```{r}
boxplot(vq$E)
```



### Differential expression testing

Fitting linear models in Limma
```{r}
fit <- lmFit(object = v, design = design)
```

Make the contrasts matrix
```{r}
contr <- makeContrasts(Stimulated - Unstimulated, levels = colnames(coef(fit)))
contr
```

Estimate contrasts for each gene
```{r}
fit2 <- contrasts.fit(fit, contr)
```

Empirical Bayes smoothing of standard errors (shrinks standard errors that are much larger or smaller than those from other genes towards the average standard error) (see https://www.degruyter.com/doi/10.2202/1544-6115.1027)
```{r}
eb <- eBayes(fit2)
```

Check out the top differential expressed genes
```{r}
topTable(eb)
```

Put all DE testing results into an object for plotting and saving
```{r}
tt2 <- topTable(eb, number = nrow(v))
tt2$Significant <- ((abs(tt2$logFC) > 1) & (tt2$adj.P.Val < 0.05))
```

How many DE genes are there?
```{r}
table(tt2$Significant)
```

### Differential expression summary plots

Volcano plot
```{r}
gg_volcano <- ggplot(tt2, aes(x = logFC, y = -log10(P.Value),
                             fill=Significant, colour=Significant)) +
    geom_point(alpha=0.5) +
    geom_vline(xintercept = c(-1, 1), linetype="dashed")
gg_volcano
```

MA plot
```{r}
gg_ma <- ggplot(tt2, aes(x = AveExpr, y = logFC, colour=Significant)) +
    geom_point(alpha=0.5) +
    geom_vline(xintercept = c(1), linetype="dashed") +
    geom_hline(yintercept = c(-1, 1), linetype="dashed")
gg_ma
```

### Significant gene plots

DE gene heatmap of DE genes
```{r, fig.height=1.5, fig.width=1}
top_de_cpm <- v$E[rownames(v$E) %in% rownames(tt2)[tt2$Significant == TRUE], ]
colnames(top_de_cpm) <- v$targets$samples

pheatmap(top_de_cpm, scale = "row", show_rownames = FALSE)
```

Plot normalised gene expression for top genes
```{r}
plot_gene <- function(gene_ids){
    
    # Get the normalised expression data for the selected genes
    dat <- v$E[rownames(v$E) %in% gene_ids, ] %>% 
        data.frame()
    colnames(dat) <- v$targets$samples
    
    # Re-shape the data into the ggplot preffered long format
    dat <- tibble::rownames_to_column(dat, var = "gene_id") %>%
        tidyr::gather(sample, value, -gene_id)
    
    # Add the group data
    dat$group <- v$targets$group[match(dat$sample, v$targets$samples)]
    
    # plot the gene expression vaues
    ggplot(data = dat, aes(x = group, y = value)) + 
        geom_point(size=3, alpha=0.5) +
        facet_wrap(~gene_id, scales = "free_y") +
        ylab("Normalised gene expression") +
        theme_bw()
        
}

my_top_genes <- rownames(tt2)[tt2$Significant == TRUE][1:6]

plot_gene(gene_ids = my_top_genes)

```

### Genome browser track plots
```{r}
txdb <- makeTxDbFromGFF(file = "reference/Homo_sapiens.GRCh38.104.chromosome.22.gtf.gz",
                        format = "gtf")

exons <- exonsBy(txdb, by = "gene")
cdss <- cdsBy(txdb, by = "gene")

track_dat <- data.frame()

selected_transcripts < my_top_genes

wiggleplotr::plotCoverage(exons = exons[my_top_genes],
                          cdss = cdss[my_top_genes], 
             ncoa7_metadata, track_data,
             heights = c(2,1), fill_palette = getGenotypePalette(),
             mean_only = FALSE, alpha = 0.5)

```

### Ontology testing
```{r}
ontology_result <- gprofiler2::gost(query = rownames(tt2)[tt2$Significant])
head(ontology_result$result)
```

Plot ontology result
```{r}
gprofiler2::gostplot(ontology_result)
```

Session information
```{r}
sessionInfo()
```

