---
title: "RNA-seq differential expression testing"
author: "Sam Buckberry"
date: "30/10/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r, message=FALSE}
library(magrittr)
library(stringr)
library(edgeR)
library(ggplot2)
library(pheatmap)
library(gprofiler2)
library(wiggleplotr)
```

Load the sample data
```{r}
mdat <- read.csv("metadata/sample_data.csv")
```

Load the gene and transcript counts data
```{r}
gene_counts <- readRDS("Effector_CD4pos_gene_counts.Rds") 
tx_counts <- readRDS("Effector_CD4pos_transcript_counts.Rds")
```

Create the DGElist object for using edgeR
```{r}
sample_ids <- gene_counts$targets %>% str_remove("_filt_fastp.bam")
y <- DGEList(counts = gene_counts$counts, samples = sample_ids)

```

Add treatment group data to the DGElist
If samples are all the correct order for the counts table and the sample metadata, you only need to do `y$samples$group <- mdat$Treatment`. However, the following code will check if all the samples are in the correct order. If you like, change the row order of the metadata by doing `mdat <- mdat[c(1,3,2,4), ]` and see if the code below corrects it. 

```{r}

# Check sample metadata and count data are in the same order, if not, fix. 
if (all(y$samples$samples == mdat$Sample)){
    y$samples$group <- mdat$Treatment
} else {
    ind <- match(y$samples$samples, mdat$Sample)
    mdat <- mdat[ind, ]
    stopifnot(mdat$Sample == y$samples$samples)
    y$samples$group <- mdat$Treatment
}

# Change to group to class factor (makes things easier later)
y$samples$group <- factor(y$samples$group,
                          levels=c("Unstimulated", "Stimulated"))


```

### Design matrix for differential testing
```{r}
design <- model.matrix(~ 0 + y$samples$group)
design
# Clean up the column names (will make things easier later)
colnames(design) <- c("Stimulated", "Unstimulated")
design
```

### Filter genes/transcripts

```{r}
keep <- filterByExpr(y, group = y$samples$group)
table(keep)

y <- y[keep, , keep.lib.sizes=FALSE]
```

### Calculate normalisation factors
```{r}
y <- calcNormFactors(y)
y$samples
```

### Estimate dispersions
```{r}
y <- estimateDisp(y)
```

### MDS Plot
How do the samples group based on the genes with the largest differences? Do they group by sample donor or by treatment group? In this type of plot, you want to look for outliers. 

```{r}
plotMDS(y, labels = y$samples$group)
```

### Boxplots

Boxplots of log2 counts for each sample
```{r, fig.height=5}
boxplot(log2(y$counts + 1), labels=y$samples$samples)
```

```{r}
y$cpm <- cpm(y)
boxplot(log2(y$cpm+1))
```

```{r}
y$cpm <- cpm(y)
boxplot(log2(y$cpm+1))
```

### Voom
```{r}
v <- voom(counts = y, design = design, plot = TRUE)
```


```{r}
boxplot(v$E)
```

If you want to normalise further using quantile normalisation
```{r}
vq <- voom(counts = y, design = design, plot = TRUE,
           normalize.method = "quantile")
```

```{r}
boxplot(vq$E)
```



### Differential expression testing

Fitting linear models in Limma
```{r}
fit <- lmFit(object = v, design = design)
head(coef(fit))
```
```{r}
contr <- makeContrasts(Stimulated - Unstimulated, levels = colnames(coef(fit)))
contr
```

Estimate contrasts for each gene
```{r}
fit2 <- contrasts.fit(fit, contr)
```

Empirical Bayes smoothing of standard errors (shrinks standard errors that are much larger or smaller than those from other genes towards the average standard error) (see https://www.degruyter.com/doi/10.2202/1544-6115.1027)
```{r}
eb <- eBayes(fit2)
```

Check out the top differential expressed genes
```{r}
topTable(eb)
```

How many DE genes are there?
```{r}
length(which(eb$adj.P.Val < 0.05))
```

Put all DE testing results into an object for plotting and saving
```{r}
tt <- topTable(eb, number = nrow(v))
tt$Significant <- abs(tt$logFC > 1) & tt$adj.P.Val < 0.05
```

### Differential expression summary plots

Volcano plot
```{r}
gg_volcano <- ggplot(tt, aes(x = logFC, y = -log10(P.Value), fill=Significant)) +
    geom_point(alpha=0.5) +
    geom_vline(xintercept = c(-1, 1), linetype="dashed") +
    
gg_volcano
```

MA plot
```{r}
gg_ma <- ggplot(tt, aes(x = AveExpr, y = logFC, fill=Significant)) +
    geom_point(alpha=0.5) +
    geom_vline(xintercept = c(1), linetype="dashed") +
    geom_hline(yintercept = c(-1, 1), linetype="dashed")
    
gg_ma
```

### Significant gene plots

DE gene heatmap of top 100 DE genes
```{r, fig.height=2, fig.width=1}
top_de_cpm <- v$E[rownames(v$E) %in% rownames(tt)[1:100], ]

pheatmap(top_de_cpm, scale = "row", show_rownames = FALSE)
```

Plot normalised gene expression for top genes
```{r}
plot_gene <- function(gene_ids){
    
    # Get the normalised expression data for the selected genes
    dat <- v$E[rownames(v$E) %in% gene_ids, ] %>% 
        data.frame()
    colnames(dat) <- v$targets$samples
    
    # Re-shape the data into the ggplot preffered long format
    dat <- tibble::rownames_to_column(dat, var = "gene_id") %>%
        tidyr::gather(sample, value, -gene_id)
    
    # Add the group data
    dat$group <- v$targets$group[match(dat$sample, v$targets$samples)]
    
    # plot the gene expression vaues
    ggplot(data = dat, aes(x = group, y = value)) + 
        geom_point(size=3, alpha=0.5) +
        facet_wrap(~gene_id, scales = "free_y") +
        ylab("Normalised gene expression") +
        theme_bw()
        
}

my_top_genes <- rownames(tt)[1:6]

plot_gene(gene_ids = my_top_genes)

```

### Genome browser track plots
```{r}
txdb <- makeTxDbFromGFF(file = "reference/Homo_sapiens.GRCh38.104.chromosome.22.gtf.gz",
                        format = "gtf")

exons <- exonsBy(txdb, by = "gene")
cdss <- cdsBy(txdb, by = "gene")

track_dat <- data.frame()

selected_transcripts < my_top_genes

wiggleplotr::plotCoverage(exons = exons[my_top_genes],
                          cdss = cdss[my_top_genes], 
             ncoa7_metadata, track_data,
             heights = c(2,1), fill_palette = getGenotypePalette(),
             mean_only = FALSE, alpha = 0.5)

```

### Ontology testing
```{r}
ontology_result <- gprofiler2::gost(query = rownames(tt)[1:100])
head(ontology_result$result)
```

Plot ontology result
```{r}
gprofiler2::gostplot(ontology_result)
```


Session information
```{r}
sessionInfo()
```

